Pull

dantoniotti opened the pull request.

dantoniotti committed:
Pull

dantoniotti committed:
Avancement exo 3

pjvilloud commented on src/main/java/com/myaudiolibrary/web/model/Album.java:
```diff
@@ -0,0 +1,51 @@
+    @JoinColumn(name = "artistId", nullable = false)
+    private Artist artist;
+
+    public Integer getAlbumId() {
+        return id;
+    }
+
+    public void setAlbumId(Integer albumId) {
+        this.id = albumId;
+    }
```
Attention à ce getter/setter qui n'est plus d'actualité !

pjvilloud commented on src/main/java/com/myaudiolibrary/web/model/Artist.java:
```diff
@@ -0,0 +1,46 @@
+    @OneToMany(mappedBy = "artist")
+    private Set<Album> albums = new HashSet<>();
+
+    public void setArtistId(Long albumId) {
+        this.id = albumId;
+    }
+
+    public Long getArtistId() {
+        return id;
+    }
```
A remplacer par le getter/setter setId et getId pour que ça fonctionne

pjvilloud reviewed (commented):
Si tu corriges ces éléments, cela devrait corriger le pb

dantoniotti committed:
Correction

dantoniotti committed:
Finalisation de l'évaluation

dantoniotti committed:
Eval REST API

dantoniotti committed:
Correction conflit d'artiste

La 409 était mal gérée.

dantoniotti committed:
Correction bug

pjvilloud commented on src/main/java/com/myaudiolibrary/web/model/Album.java:
```diff
@@ -0,0 +1,42 @@
+
+    public void setArtist(Artist artist) {
+        this.artist = artist;
+    }
```
Prend l'habitude d'ajouter les `toString`, `equals` et `hashCode`

pjvilloud commented on src/main/java/com/myaudiolibrary/web/model/Artist.java:
```diff
@@ -0,0 +1,45 @@
+    private String name;
+
+    @JsonIgnoreProperties("artist")
+    @OneToMany(cascade = CascadeType.ALL, mappedBy = "artist", orphanRemoval = true)
```
Le `CascadeType.ALL` est effectivement ici utilisable et il faut bien ajouter le `orphanRemoval` afin que la suppression de l'album soit bien persistée

Plus d'infos [ici](https://openjpa.apache.org/builds/2.4.0/apache-openjpa/docs/jpa_overview_meta_field.html#jpa_overview_meta_cascade) et [là](https://stackoverflow.com/questions/306144/jpa-cascadetype-all-does-not-delete-orphans)

pjvilloud commented on src/main/java/com/myaudiolibrary/web/model/Artist.java:
```diff
@@ -0,0 +1,45 @@
+
+    @JsonIgnoreProperties("artist")
+    @OneToMany(cascade = CascadeType.ALL, mappedBy = "artist", orphanRemoval = true)
+    private Set<Album> albums = new HashSet<>();
```
le `hashCode` est d'autant plus important dans la classe `Album` que tu utilises ici un `HashSet`. Sans le `hashCode` l'unicité n'est pas assurée.

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+@RequestMapping(value = "/artists")
+public class ArtistController {
+
+    private final String ERROR_ARTIST_NOT_FOUND = "L'artiste demandé n'existe pas !";
```
Bien les constantes

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+    private ArtistRepository artistRepository;
+
+    @RequestMapping(method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE, value = "/{idArtist}")
+    public Artist getArtiste(@PathVariable(value = "idArtist") Long idArtist) {
+        Artist artist = artistRepository.findById(idArtist).orElse(null);
+        if(artist != null) {
+            return artist;
+        } else {
+            throw new EntityNotFoundException(ERROR_ARTIST_NOT_FOUND);
+        }
+    }
```
Ce n'est pas idéal d'utiliser le `Optional` de cette manière
```suggestion
    public Optional<Artist> getArtiste(@PathVariable(value = "idArtist") Long idArtist) {
        Optional<Artist> artist = artistRepository.findById(idArtist);
        if(!artist.isEmpty()) {
            return artist;
        } 
        throw new EntityNotFoundException(ERROR_ARTIST_NOT_FOUND);
    }
```

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+
+    @RequestMapping(method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE, params = "name")
+    public Page<Artist> getArtisteByName(@RequestParam String name,
+                                         @RequestParam Integer page,
+                                         @RequestParam Integer size,
+                                         @RequestParam String sortDirection,
+                                         @RequestParam String sortProperty) {
```
Pour ce genre de paramètres n'hésite pas à mettre des valeurs par défaut

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+                                         @RequestParam Integer size,
+                                         @RequestParam String sortDirection,
+                                         @RequestParam String sortProperty) {
+        return artistRepository.findByNameContains(name, PageRequest.of(page, size, Sort.Direction.valueOf(sortDirection), sortProperty));
```
Ici, il faudrait s'assurer de la validité des paramètres pour s'assurer qu'on renvoie bien les bons codes HTTP dans notre réponse (400 et pas 500)

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+        return artistRepository.findByNameContains(name, PageRequest.of(page, size, Sort.Direction.valueOf(sortDirection), sortProperty));
+    }
+
+    @RequestMapping(method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
+    public Page<Artist> getListArtist(@RequestParam Integer page,
+                                         @RequestParam Integer size,
+                                         @RequestParam String sortDirection,
+                                         @RequestParam String sortProperty) {
+        return artistRepository.findAll(PageRequest.of(page, size, Sort.Direction.valueOf(sortDirection), sortProperty));
+    }
```
Même remarque

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+    @RequestMapping(method = RequestMethod.POST, consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseStatus(HttpStatus.CREATED)
+    public Artist createArtist(@RequestBody Artist artistCreated) {
+        if(artistRepository.findByName(artistCreated.getName()) == null) {
```
Tu peux aussi créer des méthodes `existsByName` par exemple si tu souhaites juste vérifier l'existence en base, cela sera plus rapide que de récupérer l'artiste

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+    public Artist createArtist(@RequestBody Artist artistCreated) {
+        if(artistRepository.findByName(artistCreated.getName()) == null) {
+            //
+            artistRepository.save(artistCreated);
```
Ici, tu pourrais aussi t'assurer les contraintes sur les types en BDD sont bien respectés (longueur des champs notamment)

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+            return artistCreated;
+        } else {
+            //409
+            throw new EntityExistsException("L'artiste existe déjà !");
```
A mettre en constante aussi du coup :)

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+
+    @RequestMapping(method = RequestMethod.PUT, consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE, value = "/{idArtist}")
+    public Artist modifyArtist(@RequestBody Artist artistModified, @PathVariable Long idArtist) {
+        if(artistModified.getId().equals(idArtist)) {
```
Afin d'être sûr d'éviter les `NullPointerException` je ferais plutot : 
```suggestion
        if(idArtist.equals(artistModified.getId())) {
```

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+    @RequestMapping(method = RequestMethod.PUT, consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE, value = "/{idArtist}")
+    public Artist modifyArtist(@RequestBody Artist artistModified, @PathVariable Long idArtist) {
+        if(artistModified.getId().equals(idArtist)) {
+            artistRepository.save(artistModified);
+            return artistModified;
```
```suggestion
            return artistRepository.save(artistModified);
```

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+            artistRepository.save(artistModified);
+            return artistModified;
+        } else {
+            throw new IllegalArgumentException("Les id ne correspondent pas !");
```
A mettre en constante :)

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ArtistController.java:
```diff
@@ -0,0 +1,88 @@
+    @RequestMapping(method = RequestMethod.DELETE, value = "/{idArtist}")
+    @ResponseStatus(NO_CONTENT)
+    public void deleteArtist(@PathVariable Long idArtist) {
+        Artist artist = artistRepository.findById(idArtist).orElse(null);
+        if(artist != null) {
```
Même remarque pour l'usage de l'Optional. Pour le coup tu peux utiliser aussi la méthode `existsById` pour plus d'efficacité

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/AlbumController.java:
```diff
@@ -0,0 +1,47 @@
+            albumRepository.save(albumAdded);
+            return albumAdded;
+        } else {
+            throw new EntityNotFoundException("");
```
Eviter ce genre d'écriture ! Bien mettre un message explicite (comme tu l'as fait dans le contrôleur Artist)

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/AlbumController.java:
```diff
@@ -0,0 +1,47 @@
+    @RequestMapping(method = RequestMethod.POST, consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseStatus(HttpStatus.CREATED)
+    public Album createAlbum(@RequestBody Album albumAdded) {
+        Artist artist = artistRepository.findById(albumAdded.getArtist().getId()).orElse(null);
```
cf remarque sur les Optional dans le contrôleur Artist

pjvilloud reviewed (commented):
Java POO : Acquis. Bon travail ! Pense au toString, equals et hashCode, aux constructeurs également
Java Web Services : Acquis. Très bien ! Revoir les Optional et pousser un peu plus la gestion des erreurs

dantoniotti committed:
Eval thymeleaf

dantoniotti committed:
Modification

Remplacement des objets récupérer par les repo par des optionnal.

Ajout des méthodes equals, hashcode et toString pour les classes.

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ThymeleafAudioController.java:
```diff
@@ -20,33 +25,41 @@
-        }
+    @RequestMapping(method = RequestMethod.GET, value = "/artists/{idArtist}")
+    public String getArtistById(final ModelMap model, @PathVariable Long idArtist) {
+        Optional<Artist> artist = artistRepository.findById(idArtist);
+        artist.ifPresent(value -> model.put("artist", value)); //Si l'artiste est présent alors il existe et on l'ajoute sinon 404
```
Attention ici je n'ai pas l'impression que tu gères l'erreur 404 si l'artiste n'existe pas ?

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ThymeleafAudioController.java:
```diff
@@ -59,12 +72,79 @@ public String getArtistPage(final ModelMap model, @RequestParam Integer page, @R
+            model.put("previousPage", 0);
+            model.put("currentPage", 0);
+            model.put("nextPage", 0);
+            model.put("isLastPage", true);
```
Tu vois que le code est le même que sur la méthode getArtistByName, tu peux donc extraire les instructions communes dans une méthode pour éviter la duplication de code

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ThymeleafAudioController.java:
```diff
@@ -59,12 +72,79 @@ public String getArtistPage(final ModelMap model, @RequestParam Integer page, @R
+
+    @RequestMapping(method = RequestMethod.POST, value = "/artists")
+    public String saveArtist(final ModelMap model, Artist artistCreated) {
+        //Check if exist
```
Gestion de l'erreur à ajouter, si le nom existe déjà

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ThymeleafAudioController.java:
```diff
@@ -59,12 +72,79 @@ public String getArtistPage(final ModelMap model, @RequestParam Integer page, @R
+            return new RedirectView("/thymeleaf/artists?page=0&size=10&sortProperty=name&sortDirection=ASC");
+        } else {
+            //404
+            return new RedirectView("/thymeleaf");
```
Si non trouvé, tu peux rediriger vers la page de détail de l'artiste au lieu de la page d'accueil

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ThymeleafAudioController.java:
```diff
@@ -59,12 +72,79 @@ public String getArtistPage(final ModelMap model, @RequestParam Integer page, @R
+        Optional<Album> albumOptional = albumRepository.findById(idAlbum); //On récupère l'album
+        if(artistOptional.isEmpty()) { //Si l'artiste optional est vide on redirige
+            //L'artiste n'existe pas donc on redirige vers la liste des artistes
+            return new RedirectView("/thymeleaf/artists?page=0&size=10&sortProperty=name&sortDirection=ASC");
```
Idem ici

pjvilloud commented on src/main/java/com/myaudiolibrary/web/controller/ThymeleafAudioController.java:
```diff
@@ -59,12 +72,79 @@ public String getArtistPage(final ModelMap model, @RequestParam Integer page, @R
+            return new RedirectView("/thymeleaf/artists?page=0&size=10&sortProperty=name&sortDirection=ASC");
+        }
+
+        albumOptional.ifPresent(album -> albumRepository.delete(album));
```
404 à gérer si l'album n'existe pas

pjvilloud reviewed (commented):
Thymeleaf : Maîtrisé. Travail très complet ! Quelques petites gestions d'erreur à perfectionner
